src/material-experimental/mdc-checkbox/checkbox.ts

```diff
@@ -0,0 +1,374 @@
+  @Input() labelPosition: 'before'|'after' = 'after';
+
+  /** The `name` attribute to use for the input element. */
+  @Input() name: string|null = null;
+
+  /** The `tabindex` attribute to use for the input element. */
+  @Input() tabIndex = 0;
+
+  /** The `value` attribute to use for the input element */
+  @Input() value: string;
+
+  private _uniqueId = `mat-mdc-checkbox-${++nextUniqueId}`;
+
+  /** A unique id for the checkbox. If none is supplied, it will be auto-generated. */
+  @Input() id: string = this._uniqueId;
+
+  /** Whether the checkbox is checked. */
+  @Input()
+  get checked(): boolean {
+    return this._checked;
+  }
+  set checked(checked) {
+    this._checked = coerceBooleanProperty(checked);
+  }
+  private _checked = false;
+
+  /**
+   * Whether the checkbox is indeterminate. This is also known as "mixed" mode and can be used to
+   * represent a checkbox with three states, e.g. a checkbox that represents a nested list of
+   * checkable items. Note that whenever checkbox is manually clicked, indeterminate is immediately
+   * set to false.
+   */
+  @Input()
+  get indeterminate(): boolean {
+    return this._indeterminate;
+  }
+  set indeterminate(indeterminate) {
+    this._indeterminate = coerceBooleanProperty(indeterminate);
+  }
+  private _indeterminate = false;
+
+  /** Whether the checkbox is disabled. */
+  @Input()
+  get disabled(): boolean {
+    return this._disabled;
+  }
+  set disabled(disabled) {
+    this._disabled = coerceBooleanProperty(disabled);
+    this._changeDetectorRef.markForCheck();
+  }
+  private _disabled = false;
+
+  /** Whether the checkbox is required. */
+  @Input()
+  get required(): boolean {
+    return this._required;
+  }
+  set required(required) {
+    this._required = coerceBooleanProperty(required);
+  }
+  private _required = false;
+
+  /** Event emitted when the checkbox's `checked` value changes. */
+  @Output()
+  readonly change: EventEmitter<MatCheckboxChange> = new EventEmitter<MatCheckboxChange>();
+
+  /** Event emitted when the checkbox's `indeterminate` value changes. */
+  @Output() readonly indeterminateChange: EventEmitter<boolean> = new EventEmitter<boolean>();
+
+  /** The root element for the `MDCCheckbox`. */
+  @ViewChild('checkbox') _checkbox: ElementRef<HTMLElement>;
+
+  /** The native input element. */
+  @ViewChild('nativeCheckbox') _nativeCheckbox: ElementRef<HTMLInputElement>;
+
+  /** The native label element. */
+  @ViewChild('label') _label: ElementRef<HTMLElement>;
+
+  /** Returns the unique id for the visual hidden input. */
+  get inputId(): string {
+    return `${this.id || this._uniqueId}-input`;
+  }
+
+  /** The `MDCCheckboxFoundation` instance for this checkbox. */
+  _checkboxFoundation: MDCCheckboxFoundation;
+
+  /** The set of classes that should be applied to the native input. */
+  _classes = new Map([['mdc-checkbox__native-control', true]]);
```

Comment:
We're not gaining much from using a `Map` here. Maybe we should make it into an object or an array since they don't require a polyfill.